src/diff_generate.c

```diff
@@ -989,8 +989,13 @@ static int handle_unmatched_new_item(
 	if (git_index_entry_is_conflict(nitem))
 		delta_type = GIT_DELTA_CONFLICTED;
 
-	/* update delta_type if this item is ignored */
-	else if (git_iterator_current_is_ignored(info->new_iter))
+	/* update delta_type if this item is ignored but
+	 * only if it makes a difference -- ignore checks are expensive
+	 */
```

Comment:
Hum. I'm honestly not a lot smarter after reading this comment. Could you please explain why it cannot make a difference if the below conditions are not met?

Reply:
```
	/*
	 * update delta_type if this item is ignored
	 *
	 * If the item is missing from the other iterator, or if we're not
	 * interested in ignored things, skip the check for performance
	 */
```

@romkatv I'm missing the rationale on the `!contain_oitem` check: why does it need to be added? And also, the actual skipping is done at the iterator layer (previous commit), so I'm also unsure what we're gaining by skipping again here — is it's just about sidestepping some other code in the "unmatched new item" case ?

Reply:
> I'm missing the rationale on the !contain_oitem check: why does it need to be added?

It's to avoid checking every directory in a clean repository against ignore rules. For example, when diffing index vs workdir, we might encounter these entries:

* index: foo/bar.h
* workdir: foo/

The code is trying to avoid checking "foo/" against ignore rules if it's safe to do.

It might be helpful to add a `fprintf(stderr, ...)` in `git_ignore__lookup` to see which files get checked and then run `lg2 status` in a clean (or dirty) repository. By changing the PR and seeing how it affects the results it'll be easy to see what each modification does.

> And also, the actual skipping is done at the iterator layer (previous commit)

The previous commit doesn't set `GIT_ITERATOR_SKIP_IGNORES` if `GIT_DIFF_INCLUDE_UNTRACKED` is set.

By the way, I'm positive that each successive commit has an effect because I only changed libgit2 code when it made a measurable difference on performance. However, it's possible that future commits obviated (perhaps partially) previous ones. I never went back to verify that _all_ changes are necessary to achieve the final performance.